modules/@apostrophecms/area/ui/apos/components/AposAreaWidget.vue

```diff
@@ -440,12 +459,109 @@ export default {
+
+          // If widget shifts more than tolerable don't try to track it
+          if (Math.abs(this.lastResizeTop - widgetRect.top) > this.shiftTolerance) {
+            this.stickyControlsStyles = { ...this.defaultControlsStyles };
+          } else {
+            if (this.stickyControlsStyles.position === 'fixed') {
+              this.stickyControlsStyles.right =
+                `${window.innerWidth - (widgetRect.left + widgetRect.width) + this.controlsMargin}px`;
+            }
+          }
+
+          this.lastResizeTop = widgetRect.top;
+          this.resizeTicking = false;
+        });
+      }
+    },
+    stickyControlsScroll() {
+      if (!this.scrollTicking) {
+        window.requestAnimationFrame(() => {
+          const widgetRect = this.$refs.wrapper.getBoundingClientRect();
+          const windowHeight = window.innerHeight;
+          const visibleHeight =
+            Math.min(widgetRect.bottom, windowHeight) - Math.max(widgetRect.top, 0);
+          let newStyles;
+
+          // Controls height is within 15% of the height of the widget
+          const controlsTooTallRelative = this.controlsHeight / widgetRect.height > 0.85;
+
+          // Controls are taller than widget
+          const controlsExceedWidget = widgetRect.height < this.controlsHeight;
+
+          // Repositioning in these cases feels unexpected
+          if (controlsTooTallRelative || controlsExceedWidget) {
+            return;
+          }
+
+          // Widget is under admin bar
+          if (widgetRect.top <= this.adminBarHeight) {
+
+            // Widget bottom is approaching admin bar,
+            // position controls absolutely to the bottom
+            if (
+              visibleHeight <=
+                this.controlsHeight + this.adminBarHeight + (this.controlsMargin * 3)
+            ) {
+              newStyles = {
+                position: 'absolute',
+                bottom: `${this.controlsMargin * 2}px`,
+                top: 'auto',
+                right: `${this.controlsMargin}px`
+              };
+
+            // Widget top is above admin bar,
+            // apply custom sticky position
+            } else {
+              newStyles = {
+                position: 'fixed',
+                top: `${this.controlsMargin + this.adminBarHeight}px`,
+                right: `${window.innerWidth - (widgetRect.left + widgetRect.width) + this.controlsMargin}px`
+              };
+            }
+
+          // Controls don't need positioning
+          } else {
+            newStyles = { ...this.defaultControlsStyles };
+          }
+
+          // Only update if styles changed
+          if (
+            Object.keys(newStyles).some(
+              key => this.stickyControlsStyles[key] !== newStyles[key]
+            )
+          ) {
+            this.stickyControlsStyles = newStyles;
+          }
+
+          this.scrollTicking = false;
+        });
+
+        this.scrollTicking = true;
+      }
+    },
```

Comment:
This one could be splitted in two.
this function in charge of the overall logic, and another function in charge of computing the `newStyles` and returning it.
Would make the code more readable, and remove hard to track reassignment.